src/lib/transfer/file_from_url.rb

```diff
@@ -0,0 +1,578 @@
+# encoding: utf-8
+
+# ------------------------------------------------------------------------------
+# Copyright (c) 2016 SUSE, LLC. All Rights Reserved.
+#
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of version 2 of the GNU General Public License as published by the
+# Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, contact SUSE.
+#
+# To contact SUSE about this file by physical or electronic mail, you may find
+# current contact information at www.suse.com.
+# ------------------------------------------------------------------------------
+
+require "yast"
+
+
+module Yast::Transfer
+  module FileFromUrl
+    include Yast
+
+    # lazy initialization
+    def initialize_file_from_url
+      return if @file_from_url_initialized
+      @file_from_url_initialized = true
+
+      textdomain "update"
+
+      Yast.import "URL"
+      Yast.import "FTP"
+      Yast.import "Installation"
+      Yast.import "HTTP"
+      Yast.import "StorageDevices"
+      Yast.import "TFTP"
+      Yast.import "Storage"
+      Yast.import "InstURL"
+    end
+
+    # Basename
+    # @param string path
+    # @return [String]  basename
+    def basename(filePath)
+      pathComponents = Builtins.splitstring(filePath, "/")
+      ret = Ops.get_string(
+        pathComponents,
+        Ops.subtract(Builtins.size(pathComponents), 1),
+        ""
+      )
+      ret
+    end
+
+    # Get directory name
+    # @param string path
+    # @return  [String] dirname
+    def dirname(filePath)
+      pathComponents = Builtins.splitstring(filePath, "/")
+      last = Ops.get_string(
+        pathComponents,
+        Ops.subtract(Builtins.size(pathComponents), 1),
+        ""
+      )
+      ret = Builtins.substring(
+        filePath,
+        0,
+        Ops.subtract(Builtins.size(filePath), Builtins.size(last))
+      )
+      ret
+    end
+
+    # Copy a file from a URL to a local path
+    # The URL allows autoyast-specific schemes:
+    # https://www.suse.com/documentation/sles-12/singlehtml/book_autoyast/book_autoyast.html#Commandline.ay
+    #
+    # @param scheme    [String] cifs, nfs, device, usb, http, https, ...
+    # @param host      [String]
+    # @param urlpath   [String]
+    # @param localfile [String] destination filename
+    # @param urltok    [Hash{String => String}] same url as above, but better
+    # @param destdir   [String] chroot (with crazy juggling)
+    #
+    # @return [Boolean] true on success
+    def get_file_from_url(scheme:, host:, urlpath:, localfile:,
```

Comment:
NTH: I know this was just a copy & paste but the `get_file_from_url` method is ~500 lines long and that would deserve some refactoring...